mmf/datasets/processors/text_image_processors.py

```diff
@@ -0,0 +1,157 @@
+        super().__init__(config, *args, **kwargs)
+        self.config = config
+        text_tokenizer_config = config.text_processor.tokenizer_config
+        self.text_tokenizer = AutoTokenizer.from_pretrained(
+            text_tokenizer_config.type, **text_tokenizer_config.params
+        )
+
+    def sample_traces(self, utterances, utterance_times, traces):
+        tokens, index_tokens, token_weights, token_traces = [], [], [], []
+        for idx, utterance in enumerate(utterances):
+            if self.config.remove_punctuations:
+                utterance = self.remove_punctuation(utterance)
+
+            tokens_from_utterance = self.text_tokenizer.tokenize(utterance)
+            indexed_tokens_from_utterance = self.text_tokenizer.convert_tokens_to_ids(
+                tokens_from_utterance
+            )
+            (
+                token_times,
+                token_weights_from_utterance,
+            ) = self.determine_token_weight_and_times(
+                tokens_from_utterance,
+                utterance_times[idx][0].item(),
+                utterance_times[idx][1].item(),
+            )
+            token_traces_from_utterance = self.traces_from_token(
+                token_times, traces, self.config.trace_num_samples
+            )
+
+            tokens += tokens_from_utterance
+            index_tokens += indexed_tokens_from_utterance
+            token_weights += token_weights_from_utterance
+            token_traces += token_traces_from_utterance
+        return (
+            tokens,
+            torch.IntTensor(index_tokens),
+            torch.FloatTensor(token_weights),
+            torch.FloatTensor(token_traces),
+        )
+
+    def determine_token_weight_and_times(self, tokens, start, end):
+        return (
+            torch.FloatTensor([start, end]).repeat(len(tokens), 1),
+            torch.FloatTensor([1 / len(tokens)]).repeat(len(tokens)),
+        )
+
+    def traces_from_token(self, token_times, traces, num_samples):
+        # print(f"token_times: {token_times}")
+        assert num_samples >= 2, "Must sample at least the start and end traces."
+        sampled_traces = []
+        np_traces = np.array(traces)
+        tracesx, tracesy, tracest = np_traces[:, 0], np_traces[:, 1], np_traces[:, 2]
+        for start, end in token_times:
+            interval = (end - start) / (num_samples - 1)
+            times = [start + i * interval for i in range(num_samples)]
+            sample_trace_x = np.array(
+                [
+                    interpolate.interp1d(
+                        tracest,
+                        tracesx,
+                        bounds_error=False,
+                        fill_value=traces[0][0].item()
+                        if t < traces[0][2]
+                        else traces[-1][0].item()
+                        if t > traces[-1][2]
+                        else nan,
+                    )(t)
+                    for t in times
```

Comment:
Please convert this into readable. This is very hard to read with python cuteness added :)